File: lib/robot_movement/src/4_wheel_drive/rrt_star.cpp
#include "rrt_star.h"
#include <math.h>

RRTStarPlanner::RRTStarPlanner() : RRTPlanner(), _search_radius(2.0f) {}

void RRTStarPlanner::init(float x_min, float x_max, float y_min, float y_max,
                          float step_len, float goal_sample_rate,
                          float search_radius, uint16_t iter_max) {
    RRTPlanner::init(x_min, x_max, y_min, y_max, step_len, goal_sample_rate, iter_max);
    _search_radius = search_radius;
}

float RRTStarPlanner::pathCost(int16_t idx) const {
    float cost = 0.0f;
    int16_t cur = idx;
    while (cur >= 0) {
        const int16_t p = _nodes[cur].parent;
        if (p < 0) {
            break;
        }
        cost += dist(_nodes[cur].x, _nodes[cur].y, _nodes[p].x, _nodes[p].y);
        cur = p;
    }
    return cost;
}

void RRTStarPlanner::nearNeighbors(float x, float y, int16_t *out, uint8_t &out_n) const {
    out_n = 0;
    const float n = (float)_node_count + 1.0f;
    float radius = _search_radius * sqrtf(logf(n) / n);
    if (radius > _step_len * 5.0f) {
        radius = _step_len * 5.0f;
    }
    if (radius < _step_len) {
        radius = _step_len;
    }
    for (uint8_t i = 0; i < _node_count && out_n < 16; i++) {
        if (dist(_nodes[i].x, _nodes[i].y, x, y) <= radius) {
            if (collisionFree(_nodes[i].x, _nodes[i].y, x, y)) {
                out[out_n++] = (int16_t)i;
            }
        }
    }
}

int16_t RRTStarPlanner::chooseParent(float x, float y, const int16_t *near, uint8_t near_n) const {
    int16_t best = nearest(x, y);
    float best_cost = pathCost(best) + dist(_nodes[best].x, _nodes[best].y, x, y);
    for (uint8_t i = 0; i < near_n; i++) {
        const int16_t ni = near[i];
        const float c = pathCost(ni) + dist(_nodes[ni].x, _nodes[ni].y, x, y);
        if (c < best_cost) {
            best_cost = c;
            best = ni;
        }
    }
    return best;
}

void RRTStarPlanner::rewire(int16_t new_i, const int16_t *near, uint8_t near_n) {
    for (uint8_t i = 0; i < near_n; i++) {
        const int16_t ni = near[i];
        if (ni == new_i || ni == _nodes[new_i].parent) {
            continue;
        }
        const float through_new =
            pathCost(new_i) + dist(_nodes[new_i].x, _nodes[new_i].y, _nodes[ni].x, _nodes[ni].y);
        if (through_new < pathCost(ni)) {
            if (collisionFree(_nodes[new_i].x, _nodes[new_i].y, _nodes[ni].x, _nodes[ni].y)) {
                _nodes[ni].parent = new_i;
            }
        }
    }
}

int16_t RRTStarPlanner::bestGoalParent() const {
    int16_t best = -1;
    float best_cost = 1e30f;
    for (uint8_t i = 0; i < _node_count; i++) {
        const float d = dist(_nodes[i].x, _nodes[i].y, _goal_x, _goal_y);
        if (d <= _step_len && collisionFree(_nodes[i].x, _nodes[i].y, _goal_x, _goal_y)) {
            const float c = pathCost((int16_t)i) + d;
            if (c < best_cost) {
                best_cost = c;
                best = (int16_t)i;
            }
        }
    }
    return best;
}

bool RRTStarPlanner::plan(float start_x, float start_y, float goal_x, float goal_y) {
    _path_len = 0;
    _node_count = 0;
    _goal_x = goal_x;
    _goal_y = goal_y;

    if (insideObstacle(start_x, start_y) || insideObstacle(goal_x, goal_y)) {
        return false;
    }

    _nodes[0].x = start_x;
    _nodes[0].y = start_y;
    _nodes[0].parent = -1;
    _node_count = 1;

    for (uint16_t it = 0; it < _iter_max; it++) {
        if (_node_count >= RM_RRT_MAX_NODES) {
            break;
        }
        float rx, ry;
        sample(rx, ry);
        const int16_t near_i = nearest(rx, ry);
        float nx, ny;
        if (!steer(_nodes[near_i].x, _nodes[near_i].y, rx, ry, nx, ny)) {
            continue;
        }
        if (!collisionFree(_nodes[near_i].x, _nodes[near_i].y, nx, ny)) {
            continue;
        }

        int16_t near_list[16];
        uint8_t near_n = 0;
        nearNeighbors(nx, ny, near_list, near_n);
        const int16_t parent = chooseParent(nx, ny, near_list, near_n);
        if (!collisionFree(_nodes[parent].x, _nodes[parent].y, nx, ny)) {
            continue;
        }

        const int16_t new_i = (int16_t)_node_count;
        _nodes[new_i].x = nx;
        _nodes[new_i].y = ny;
        _nodes[new_i].parent = parent;
        _node_count++;
        rewire(new_i, near_list, near_n);
    }

    const int16_t goal_parent = bestGoalParent();
    if (goal_parent < 0) {
        return false;
    }
    if (_node_count < RM_RRT_MAX_NODES) {
        _nodes[_node_count].x = goal_x;
        _nodes[_node_count].y = goal_y;
        _nodes[_node_count].parent = goal_parent;
        extractPath((int16_t)_node_count);
        _node_count++;
    } else {
        extractPath(goal_parent);
    }
    return _path_len > 0;
}

File: python/path_planning/rrt_star.py
import math
import random

from .env import ContinuousMap
from .rrt import Node


class RRTStar:
    def __init__(
        self,
        start,
        goal,
        step_len=0.5,
        goal_sample_rate=0.1,
        search_radius=5.0,
        iter_max=5000,
        env_map=None,
    ):
        self.start = Node(start[0], start[1])
        self.goal = Node(goal[0], goal[1])
        self.step_len = step_len
        self.goal_sample_rate = goal_sample_rate
        self.search_radius = search_radius
        self.iter_max = iter_max
        self.env = env_map if env_map is not None else ContinuousMap()
        self.vertex = [self.start]
        self.path = []

    def plan(self):
        for _ in range(self.iter_max):
            rnd = self._random_node()
            nearest = self._nearest(rnd)
            new = self._steer(nearest, rnd)
            if new is None:
                continue
            if self.env.is_collision(nearest.x, nearest.y, new.x, new.y):
                continue

            near_indices = self._near_neighbors(new)
            self._choose_parent(new, near_indices)
            self.vertex.append(new)
            self._rewire(new, near_indices)

        goal_idx = self._best_goal_parent()
        if goal_idx is None:
            return None
        self.path = self._extract_path(self.vertex[goal_idx])
        return self.path

    def _random_node(self):
        if random.random() < self.goal_sample_rate:
            return Node(self.goal.x, self.goal.y)
        x_min, x_max = self.env.x_range
        y_min, y_max = self.env.y_range
        return Node(
            random.uniform(x_min + 0.5, x_max - 0.5),
            random.uniform(y_min + 0.5, y_max - 0.5),
        )

    def _nearest(self, node):
        return min(
            self.vertex,
            key=lambda n: math.hypot(n.x - node.x, n.y - node.y),
        )

    def _steer(self, from_node, to_node):
        dist = math.hypot(to_node.x - from_node.x, to_node.y - from_node.y)
        if dist == 0:
            return None
        step = min(self.step_len, dist)
        theta = math.atan2(to_node.y - from_node.y, to_node.x - from_node.x)
        return Node(
            from_node.x + step * math.cos(theta),
            from_node.y + step * math.sin(theta),
        )

    def _near_neighbors(self, node):
        n = len(self.vertex) + 1
        r = min(self.search_radius * math.sqrt(math.log(n) / n), self.step_len * 5)
        indices = []
        for i, v in enumerate(self.vertex):
            if math.hypot(v.x - node.x, v.y - node.y) <= r:
                if not self.env.is_collision(v.x, v.y, node.x, node.y):
                    indices.append(i)
        return indices

    def _cost(self, node):
        cost = 0.0
        current = node
        while current.parent is not None:
            cost += math.hypot(
                current.x - current.parent.x, current.y - current.parent.y
            )
            current = current.parent
        return cost

    def _choose_parent(self, new, near_indices):
        if not near_indices:
            nearest = self._nearest(new)
            new.parent = nearest
            return
        best = None
        best_cost = math.inf
        for i in near_indices:
            candidate = self.vertex[i]
            cost = self._cost(candidate) + math.hypot(
                candidate.x - new.x, candidate.y - new.y
            )
            if cost < best_cost:
                best_cost = cost
                best = candidate
        new.parent = best

    def _rewire(self, new, near_indices):
        for i in near_indices:
            neighbor = self.vertex[i]
            new_cost = self._cost(new) + math.hypot(
                neighbor.x - new.x, neighbor.y - new.y
            )
            if new_cost < self._cost(neighbor):
                if not self.env.is_collision(new.x, new.y, neighbor.x, neighbor.y):
                    neighbor.parent = new

    def _best_goal_parent(self):
        candidates = []
        for i, v in enumerate(self.vertex):
            dist = math.hypot(v.x - self.goal.x, v.y - self.goal.y)
            if dist <= self.step_len:
                if not self.env.is_collision(v.x, v.y, self.goal.x, self.goal.y):
                    candidates.append((self._cost(v) + dist, i))
        if not candidates:
            return None
        candidates.sort(key=lambda c: c[0])
        return candidates[0][1]

    def _extract_path(self, end_node):
        path = [(self.goal.x, self.goal.y)]
        node = end_node
        while node is not None:
            path.append((node.x, node.y))
            node = node.parent
        path.reverse()
        return path
